File: 22. Generate Parentheses Medium/solution.ts
function generateParenthesis(n: number): string[] {
    let result: string[] = [];
    recGenerator("", 0, 0, n, result);
    return result;
}

function recGenerator(currentString: string, opened: number, closed: number, n: number, result: string[]): void {
    if (currentString.length == 2*n) {
        result.push(currentString);
        return;
    }
    if (opened < n) {
        recGenerator(currentString + "(", opened + 1, closed, n, result);
    }
    if (closed < opened) {
        recGenerator(currentString + ")", opened, closed + 1, n, result);
    }
}

console.log(generateParenthesis(3))
console.log(generateParenthesis(1))

File: 22. Generate Parentheses Medium/solution.py
def generate_parenthesis(n: int) -> list[str]:
    result = []
    rec_generator("", 0, 0, n, result)
    return result

def rec_generator(current_string: str, opened: int, closed: int, n: int, result: list) -> None:
    if len(current_string) == 2 * n:
        result.append(current_string)
        return
    if opened < n:
        rec_generator(current_string + '(', opened + 1, closed, n, result)
    if closed < opened:
        rec_generator(current_string + ')', opened, closed + 1, n, result)


print(generate_parenthesis(3))
print(generate_parenthesis(1))